graphlearn_torch/python/distributed/dist_loader.py

```diff
@@ -122,105 +124,139 @@ def __init__(self,
-        self._channel.pin_memory()
 
-      self._mp_producer = DistMpSamplingProducer(
-        self.data, self.input_data, self.sampling_config,
-        self.worker_options, self._channel
-      )
-      self._mp_producer.init()
-
-    elif isinstance(self.worker_options, RemoteDistSamplingWorkerOptions):
+    if self._is_remote_worker:
       if not current_ctx.is_client():
-        raise RuntimeError(f"'{self.__class__.__name__}': `DistNeighborLoader` "
-                           f"must be used on a client worker process.")
+        raise RuntimeError(
+          f"'{self.__class__.__name__}': `DistNeighborLoader` "
+          f"must be used on a client worker process."
+        )
+      self._num_expected = float(
+        'inf'
+      )  # for remote worker, end of epoch is determined by server
 
       # Launch remote sampling workers
-      self._worker_mode = 'remote'
       self._with_channel = True
-      self.worker_options._set_worker_ranks(current_ctx)
 
-      self._server_rank = self.worker_options.server_rank
-      if self._server_rank is None:
-        self._server_rank = current_ctx.rank % current_ctx.num_servers()
+      self._server_rank_list = self.worker_options.server_rank \
+        if isinstance(self.worker_options.server_rank, List) else [self.worker_options.server_rank]
+      self._input_data_list = self.input_data if isinstance(
+        self.input_data, List
+      ) else [self.input_data]
+
+      self._input_type = self._input_data_list[0].input_type
 
       self.num_data_partitions, self.data_partition_idx, ntypes, etypes = \
-        request_server(self._server_rank, DistServer.get_dataset_meta)
+        request_server(self._server_rank_list[0], DistServer.get_dataset_meta)
       self._set_ntypes_and_etypes(ntypes, etypes)
 
-      self._producer_id = request_server(
-        self._server_rank,
-        DistServer.create_sampling_producer,
-        self.input_data.to(torch.device('cpu')),
-        self.sampling_config,
-        self.worker_options
-      )
-
+      self._producer_id_list = []
+      futures = []
+      for input_data in self._input_data_list:
+        if not isinstance(input_data, RemoteSamplerInput):
+          input_data = input_data.to(torch.device('cpu'))
+
+      with concurrent.futures.ThreadPoolExecutor() as executor:
+        futures = [executor.submit(request_server, server_rank, DistServer.create_sampling_producer, input_data, self.sampling_config, self.worker_options) \
+                         for server_rank, input_data in zip(self._server_rank_list, self._input_data_list)]
+        
+      for future in futures:
+        producer_id = future.result()
+        self._producer_id_list.append(producer_id)
       self._channel = RemoteReceivingChannel(
-        self._server_rank, self._producer_id,
-        self._num_expected, self.worker_options.prefetch_size
+        self._server_rank_list, self._producer_id_list,
+        self.worker_options.prefetch_size
       )
-
     else:
```

Comment:
Test the examples of the other two cases

Reply:
have run examples/distributed/dist_train_sage_supervised.py for MpDistSamplingWorkerOptions